### PR TITLE
1242 idsite pre and post handlers

### LINKED ISSUE #1242: Pre and Post Handlers are not invoked when IDSite is enabled
Pre/Post Login/Register Handlers are not invoked when IDSite is enabled.

### PR DESCRIPTION
**Pull request by mrioan**
Resolves #1242

If 1.4.0 gets release before this code is approved then this PR should be updated to target 1.4.x

### CODE REVIEW

**Comment by mrioan** [comment]
This branch is already being used by a customer who confirms that things are working fine for him:

See support ticket 10950

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/CallbackController.java [hunk 1]
```diff
@@ -106,6 +107,12 @@ protected ViewModel onAuthentication(HttpServletRequest request,
 
         publish(new DefaultSuccessfulAuthenticationRequestEvent(request, response, null, authcResult));
 
+        if (postLoginHandler != null) {
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/CallbackController.java]: Is there any reason not to collapse this to:

`if (postLoginHandler != null && !postLoginHandler.handle(request, response, result.getAccount()) {`

?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java [hunk 2]
```diff
@@ -65,13 +68,22 @@ public void init() {
         Assert.notNull(callbackUri, "callbackUri must be configured.");
         Assert.notNull(idSiteOrganizationResolver, "idSiteOrganizationResolver must be configured.");
         Assert.notNull(alreadyLoggedInUri, "alreadyLoggedInUri must be configured.");
+        Assert.isTrue(!(preRegisterHandler != null && preLoginHandler != null), "This IDSite controller must contain only one of preRegisterHandler and preLoginHandler");
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java]: Why can you not have both a pre-login handler and a pre-register handler? You can when not using id site.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteResultController.java [hunk 3]
```diff
@@ -126,7 +127,12 @@ private ViewModel onRegistration(final HttpServletRequest request, final HttpSer
             AuthenticationResult authcResult = new TransientAuthenticationResult(account);
             saveResult(request, response, authcResult);
         }
-        // else - do we need to do anything else?
+
+        if (postRegisterHandler != null) {
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteResultController.java]: may be more work than is worth it, but I wonder if we should have default handlers that return false so we don't need the null checks? NOT high prio...

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/CallbackController.java (hunk 1 above)
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/CallbackController.java]: Yeah, right... Changed!

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java (hunk 2 above)
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java]: @dogeared the truth is that this `IDSiteController` is instantiated twice (one for `Login` and another for `Registration`; actually I think there `ForgotPassword` and others create more instances as well). So for the sake of speed (a customer need a quick a fix) I did not do what I should have done: creating specific new classes for each case. Do you want me to code it the proper way?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteResultController.java (hunk 3 above)
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteResultController.java]: Makes sense but, as you said, this change will impact several places and there will not be a major benefit now...

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java (hunk 2 above)
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java]: I am thinking that I should really create those classes: with the current status a `RegisterController` could have a `LoginHandler` attached....

Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/IdSiteController.java]: oh, I see. I understand we are in a crunch for a client. If you want to release as-is, that's fine. Let's just create an issue for the next release to refactor and clean it up a bit.

**Comment by dogeared** [comment]
After CI, please merge

**Pull request merged**